### PR TITLE
Cache mpTelemetry routes

### PR DESCRIPTION
**Pull request by jdmcclur**
Cache routes based on Jared's routekey idea

### CODE REVIEW

**Comment by jdmcclur** [comment]
#build

**Comment by jdmcclur** [comment]
The personal build has an odd failure, but I cant see how it's related to this change. I have done some manually testing with this change, and it works well. Perf is 4 to 5% better and items go out of the Hashmap when an application stops (on the next GC).

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java [hunk 1]
```diff
@@ -44,12 +48,130 @@ public class TelemetryContainerFilter implements ContainerRequestFilter, Contain
...
+            }
+            if (!restMethodRef.equals(other.restMethodRef)) {
+                return false;
+            }
+            return true;
+        }
+    }
+
+    private static class RestRouteKeyWeakReference<T> extends WeakReference<T> {
+        private final int hash;
```
Comment (jhanders34) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java]:
```suggestion
```

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java [hunk 2]
```diff
@@ -44,12 +48,130 @@ public class TelemetryContainerFilter implements ContainerRequestFilter, Contain
...
+        private final RestRouteKey owningKey;
+
+        RestRouteKeyWeakReference(T referent, RestRouteKey owningKey) {
+            super(referent);
+            this.owningKey = owningKey;
+            hash = referent.hashCode();
```
Comment (jhanders34) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java]:
```suggestion
```

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java [hunk 3]
```diff
@@ -44,12 +48,130 @@ public class TelemetryContainerFilter implements ContainerRequestFilter, Contain
...
+        }
+
+        RestRouteKeyWeakReference(T referent, RestRouteKey owningKey,
+                                  ReferenceQueue<T> referenceQueue) {
+            super(referent, referenceQueue);
+            this.owningKey = owningKey;
+            hash = referent.hashCode();
```
Comment (jhanders34) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java]:
```suggestion
```

File: dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java [hunk 4]
```diff
@@ -44,12 +48,130 @@ public class TelemetryContainerFilter implements ContainerRequestFilter, Contain
...
+
+        RestRouteKey getOwningKey() {
+            return owningKey;
+        }
+
+        @Override
+        public int hashCode() {
+            return hash;
+        }
+
```
Comment (jhanders34) [inline dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/rest/TelemetryContainerFilter.java]:
```suggestion
```

**Review by jhanders34 — APPROVED** [review]
Extra code that isn't needed is recommended to be removed.

**Comment by jdmcclur** [comment]
#build #libby

**Review by Yas-Franco — APPROVED** [review]
Looks good! Thanks

**Pull request merged**